src/modules/aggsig/main_impl.h

```diff
@@ -0,0 +1,339 @@
+    size_t i;
+    secp256k1_scalar s;
+    secp256k1_ge final;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(aggctx != NULL);
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(partial != NULL);
+    (void) ctx;
+
+    if (n_sigs != aggctx->n_sigs) {
+        return 0;
+    }
+
+    secp256k1_scalar_set_int(&s, 0);
+    for (i = 0; i < n_sigs; i++) {
+        secp256k1_scalar tmp;
+        int overflow;
+        secp256k1_scalar_set_b32(&tmp, partial[i].data, &overflow);
+        if (overflow) {
+            return 0;
+        }
+        secp256k1_scalar_add(&s, &s, &tmp);
+    }
+
+    /* If we need to negate the public nonce, everyone will
+     * have negated their secret nonces in the previous step. */
+    if (!secp256k1_gej_has_quad_y_var(&aggctx->pubnonce_sum)) {
+        secp256k1_gej_neg(&aggctx->pubnonce_sum, &aggctx->pubnonce_sum);
+    }
+
+    secp256k1_scalar_get_b32(sig64, &s);
+    secp256k1_ge_set_gej(&final, &aggctx->pubnonce_sum);
+    secp256k1_fe_normalize_var(&final.x);
+    secp256k1_fe_get_b32(sig64 + 32, &final.x);
+    return 1;
+}
+
+SECP256K1_INLINE static void secp256k1_aggsig_endo_split(secp256k1_scalar *s1, secp256k1_scalar *s2, secp256k1_gej *p1, secp256k1_gej *p2) {
+    secp256k1_scalar tmp = *s1;
+    secp256k1_scalar_split_lambda(s1, s2, &tmp);
+    secp256k1_gej_mul_lambda(p2, p1);
+
+    if (secp256k1_scalar_is_high(s1)) {
+        secp256k1_scalar_negate(s1, s1);
+        secp256k1_gej_neg(p1, p1);
+    }
+    if (secp256k1_scalar_is_high(s2)) {
+        secp256k1_scalar_negate(s2, s2);
+        secp256k1_gej_neg(p2, p2);
+    }
+}
+
+int secp256k1_aggsig_verify(const secp256k1_context* ctx, const unsigned char *sig64, const unsigned char *msg32, const secp256k1_pubkey *pubkeys, size_t n_pubkeys) {
+    secp256k1_gej pt[SECP256K1_ECMULT_MULTI_MAX_N];
+    secp256k1_scalar sc[SECP256K1_ECMULT_MULTI_MAX_N];
+    secp256k1_gej pk_sum;
+    secp256k1_fe fe_tmp;
+    secp256k1_ge ge_tmp;
+    size_t i;
+    size_t offset;
+    int overflow;
+    unsigned char prehash[32];
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(pubkeys != NULL);
+    (void) ctx;
+
+    if (n_pubkeys == 0) {
+        return 0;
+    }
+
+    /* Compute sum -sG + R + e_i*P_i */
+    secp256k1_gej_set_infinity(&pk_sum);
+    secp256k1_scalar_set_b32(&sc[0], sig64, &overflow);
+    if (overflow) {
+        return 0;
+    }
+    secp256k1_scalar_negate(&sc[0], &sc[0]);
+    secp256k1_gej_set_ge(&pt[0], &secp256k1_ge_const_g);
+
+    if (!secp256k1_fe_set_b32(&fe_tmp, sig64 + 32)) {
+        return 0;
+    }
+    secp256k1_scalar_set_int(&sc[1], 1);
+    secp256k1_ge_set_xquad(&ge_tmp, &fe_tmp);
+    secp256k1_gej_set_ge(&pt[1], &ge_tmp);
+    secp256k1_compute_prehash(ctx, prehash, pubkeys, n_pubkeys, &ge_tmp, msg32);
+
+    i = 0;
+#ifdef USE_ENDOMORPHISM
+    secp256k1_aggsig_endo_split(&sc[0], &sc[2], &pt[0], &pt[2]);
```

Comment:
It would be nice if the endomorphism splitting code was in secp256k1_ecmult_multi (like it is in secp256k1_ecmult). That way the aggsig code (and any other multi-mul users) can be free of ifdefs.